### PR TITLE
gpt: support LoongArch 64-bit

### PR DESCRIPTION
**Pull request by xen0n**
This PR adds LoongArch partition types to the Discoverable Partitions Specification, in preparation of forthcoming LoongArch support. All known LoongArch systems are expected to support ACPI and UEFI.

FYI, LoongArch is a new RISC architecture developed by the Loongson Corporation, after switching away from MIPS. You could find the currently available official documentation [here](https://github.com/loongson/LoongArch-Documentation), and toolchain [here](https://github.com/loongson/build-tools) (binary-only for now). I'm not Loongson employee but am working with Loongson guys at leisure.

### CODE REVIEW

File: docs/DISCOVERABLE_PARTITIONS.md [hunk 1]
```diff
@@ -221,13 +225,13 @@ We are not. `/etc/fstab` always overrides automatic discovery and is indeed
 mentioned in the specifications.  We are simply trying to make the boot and
 installation processes of Linux a bit more robust and self-descriptive.
 
-### Why did you only define the root partition for x86, x86-64, ARM, ARM64, ia64, riscv32, riscv64?
+### Why did you only define the root partition for x86, x86-64, ARM, ARM64, ia64, loongarch64, riscv32, riscv64?
```
Comment (keszybz) [inline docs/DISCOVERABLE_PARTITIONS.md]: I think this should be capitalized the same as in the text.

**Review by keszybz — COMMENTED** [review]
Looks reasonable.

> Since the academic level of the original authors is extraordinary high, words and phrases in the original documents are so profound that the translators with limited knowledge could not express them accurately even if they tried their best.

That's some serious stuff.

File: docs/DISCOVERABLE_PARTITIONS.md (hunk 1 above)
Comment (xen0n) [reply docs/DISCOVERABLE_PARTITIONS.md]: For this part I followed the "majority" in that list; but actually I think the sentence might be too long for a title. Would a question like "Why did you only define the root partition for these listed architectures?" be reasonable enough?

**Comment by xen0n** [comment]
> > Since the academic level of the original authors is extraordinary high, words and phrases in the original documents are so profound that the translators with limited knowledge could not express them accurately even if they tried their best.
> 
> That's some serious stuff.

And that's probably what you would say, if you are faced with translating material from several PhD's while you don't have that degree ;-)

**Comment by xen0n** [comment]
Besides, the canonical name of the architecture is long (`loongarch64`), so a cosmetic change is needed as you can see. Because still other constants need to be defined for full-fledged support, would contracting the name to `LA64` be okay? I've been told by multiple people that `la64` is visually similar to `ia64` which is not desired, so I didn't use that. If we choose the full name as in this PR, expect things like `ARCHITECTURE_LOONGARCH64` to pop up in some following PRs; we might have to allow some non-uniformity there, sadly.

**Comment by poettering** [comment]
hmm, usually an addition to `src/basic/architecture.h` and `.c` would come first?

is this new arch already in the linux kernel?

I'd probably postpone this until the kernel/toolchain stuff is merged. systemd is kind the third step only I'd say were this has to land.

File: docs/DISCOVERABLE_PARTITIONS.md (hunk 1 above)
Comment (keszybz) [reply docs/DISCOVERABLE_PARTITIONS.md]: "Why did you only define the root partition for these listed architectures?" — yes please.

**Comment by keszybz** [comment]
> Besides, the canonical name of the architecture is long (`loongarch64`), so a cosmetic change is needed as you can see. Because still other constants need to be defined for full-fledged support, would contracting the name to `LA64` be okay? I've been told by multiple people that `la64` is visually similar to `ia64` which is not desired, so I didn't use that. If we choose the full name as in this PR, expect things like `ARCHITECTURE_LOONGARCH64` to pop up in some following PRs; we might have to allow some non-uniformity there, sadly.

I think both LoongArch64 and `ARCHITECTURE_LOONGARCH64` is OK. `la64` is hard to distinguish from `IA64`, and I hope this doesn't get used. Instead, I would hope people start using Loong for it. (Do you pronounce it /ˈluːŋ/, like "loon" with a "g" at the end?)

> I'd probably postpone this until the kernel/toolchain stuff is merged. systemd is kind the third step only I'd say were this has to land.

I would say it's fine to merge this here early. It allows people to reserve some uuids. Even if the architecture were to disappear off the face of Earth tomorrow, this doesn't cause us any problems. We could simply delete the few lines if the defines are not used anywhere. But including this here we are not making any promises.

OTOH, when you're trying to bring up a new architecture, it helps to be able to push some preparatory bits early to get the ball rolling. We could accept patches that don't touch code first.

File: docs/DISCOVERABLE_PARTITIONS.md (hunk 1 above)
Comment (keszybz) [reply docs/DISCOVERABLE_PARTITIONS.md]: Can you update the PR like this please?

Comment (xen0n) [reply docs/DISCOVERABLE_PARTITIONS.md]: Oh sure. This got lost in my GitHub inbox :new_moon_with_face:

**Comment by xen0n** [comment]
Rebased and fixed that doc title line.

**Comment by xen0n** [comment]
> > Besides, the canonical name of the architecture is long (`loongarch64`), so a cosmetic change is needed as you can see. Because still other constants need to be defined for full-fledged support, would contracting the name to `LA64` be okay? I've been told by multiple people that `la64` is visually similar to `ia64` which is not desired, so I didn't use that. If we choose the full name as in this PR, expect things like `ARCHITECTURE_LOONGARCH64` to pop up in some following PRs; we might have to allow some non-uniformity there, sadly.
> 
> I think both LoongArch64 and `ARCHITECTURE_LOONGARCH64` is OK. `la64` is hard to distinguish from `IA64`, and I hope this doesn't get used. Instead, I would hope people start using Loong for it. (Do you pronounce it /ˈluːŋ/, like "loon" with a "g" at the end?)

Yes, sure. The common practice of Loongson employees doing the porting work seems to prefer the fully spelled out "LoongArch", but something like "loong" is certainly possible. Though it seems they have to go through some tedious internal discussions to be permitted to use "loong", like the episode in golang/go#46229. :thinking:

As for the pronunciation, /ˈluːŋ/ is very close and acceptable for a non-native speaker! Because "loong" really is just romanized "龙", natives pronounce it like "龙 arch" /'lʊŋ ɑɹk/, but /ˈluːŋ/ is 100% understandable.

**Comment by keszybz** [comment]
I'll merge this, see justification in https://github.com/systemd/systemd/pull/20087#issuecomment-872240618.

**Pull request merged**